rfc/0010-cbft.md

```diff
@@ -0,0 +1,566 @@
+
+For both bucket and cluster level, the approach is the same. Right now only Bucket level is defined in this RFC.
+
+```java
+bucket.query(SearchQuery[, SearchParams]);
+```
+
+All query types described below must implement the `SearchQuery` marker interface or base class (including compound queries since they can be nested too).
+
+Here is an example of a String query with some search params:
+
+```java
+bucket.query(new StringQuery("description:awesome").boost(2), new SearchParams().limit(5))
+```
+
+If idiomatic by the language, it makes sense to provide static factory methods (or similar helpers) to guide the user in the right direction since there are many different query options available. So for example the `SearchQuery` can contain factory methods in a form like this:
+
+```java
+bucket.query(SearchQuery.string("+water +abv:>10"), new SearchParams().limit(5))
+```
+
+This removes cognitive load, since the user only needs to remember "SearchQuery" and will have all supported query methods available immediately.
+
+## Request Query Options
+
+### Search Request Params
+All options described in this section are not bound to a specific search option, but rather treated as a sparate param set as an optional argument. They affect the query as a whole, independent of the actual search method (including compound queries) used.
+
+They are part of the `SearchParams` object which can be used optionally by the user. If not used, default settings as noted should be applied on every request (for example to make sure the server timeout is always passed).
+
+`new SearchParams().limit(1).skip(5).explain()`
+
+#### Limit
+The `limit` option limits the number of matches returned from the complete result set. The actual HTTP query field is called `size`, but to match it with N1QL and Views it makes sense to name it similarly.
+
+| Builder Name | Request Path | Wire Type | Default | Constraints |
+| ------------ | ------------ | --------- | ------- | ----------- |
+| `limit(int limit)` | `size` | JSON Number | - | >  0, if not set no size is used (no default value) |
+
+**Examples:** `.limit(10)`
+
+#### Skip
+The `skip` option indicates how many matches are skipped on the result set before starting to return the matches. The actual HTTP query field is called `from`, but to match it with N1QL and Views it makes sense to name it similarly.
+
+| Builder Name | Request Path | Wire Type | Default | Constraints |
+| ------------ | ------------ | --------- | ------- | ----------- |
+| `skip(int skip)` | `from` | JSON Number | - | >  0, if not set no size is used (no default value) |
+
+**Examples:** `.skip(5)`
+
+#### Explain
+The boolean `explain` field triggers inclusion of additional search result score explanations.
+
+| Builder Name | Request Path | Wire Type | Default | Constraints |
+| ------------ | ------------ | --------- | ------- | ----------- |
+| `explain()` | `explain` | JSON Boolean | - | -|
```

Comment:
chaining in this method when building the query would mean you want `"explain": true` in your request, so the assumed boolean value is `true`.